tablecodec/tablecodec.go

```diff
@@ -599,47 +601,116 @@ func CutIndexKeyNew(key kv.Key, length int) (values [][]byte, b []byte, err erro
-	values, b, err := CutIndexKeyNew(key, colsLen)
+func handleExists(hdStatus HandleStatus) bool {
+	return hdStatus != HandleNotExists
+}
+
+// reEncodeHandleByStatus first decode the value into a int or uint decided by the hdStatus, then encode it so that it can
+// be properly decoded.
+func reEncodeHandleByStatus(value []byte, hdStatus HandleStatus) ([]byte, error) {
+	handle, err := DecodeIndexValueAsHandle(value)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
-	if len(b) > 0 {
-		if pkStatus != PrimaryKeyNotExists {
-			values = append(values, b)
+	var handleDatum types.Datum
+	if hdStatus == HandleIsUnsigned {
+		handleDatum = types.NewUintDatum(uint64(handle))
+	} else {
+		handleDatum = types.NewIntDatum(handle)
+	}
+	handleBytes := make([]byte, 0, 8)
+	handleBytes, err = codec.EncodeValue(nil, handleBytes, handleDatum)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return handleBytes, nil
+}
+
+func decodeIndexKvNewCollation(key, value []byte, colsLen int, hdStatus HandleStatus) ([][]byte, error) {
+	resultValues, err := cutIndexValue(value, colsLen)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	tailLen := value[0]
+	if tailLen <= 1 {
+		// In non-unique index.
+		if handleExists(hdStatus) {
+			_, b, err := CutIndexKeyNew(key, colsLen)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			resultValues = append(resultValues, b)
 		}
-	} else if pkStatus != PrimaryKeyNotExists {
-		handle, err := DecodeIndexValueAsHandle(value)
-		if err != nil {
-			return nil, errors.Trace(err)
+	} else {
+		// In unique index.
+		if handleExists(hdStatus) {
+			handleBytes, err := reEncodeHandleByStatus(value[len(value)-int(tailLen):], hdStatus)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			resultValues = append(resultValues, handleBytes)
 		}
-		var handleDatum types.Datum
-		if pkStatus == PrimaryKeyIsUnsigned {
-			handleDatum = types.NewUintDatum(uint64(handle))
-		} else {
-			handleDatum = types.NewIntDatum(handle)
+	}
+	return resultValues, nil
+}
+
+func decodeIndexKvOldCollation(key, value []byte, colsLen int, hdStatus HandleStatus) ([][]byte, error) {
+	resultValues, b, err := CutIndexKeyNew(key, colsLen)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(b) > 0 {
+		// non-unique index
+		if handleExists(hdStatus) {
+			resultValues = append(resultValues, b)
 		}
-		handleBytes := make([]byte, 0, 8)
-		handleBytes, err = codec.EncodeValue(nil, handleBytes, handleDatum)
+	} else if handleExists(hdStatus) {
+		// unique index
+		handleBytes, err := reEncodeHandleByStatus(value, hdStatus)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}
-		values = append(values, handleBytes)
+		resultValues = append(resultValues, handleBytes)
+	}
+	return resultValues, nil
+}
+
+// DecodeIndexKV uses to decode index key values.
+func DecodeIndexKV(key, value []byte, colsLen int, hdStatus HandleStatus) ([][]byte, error) {
+	if len(value) > MaxOldEncodeValueLen {
```

Comment:
Is it possible that `len(indexedColumnsValue) <= MaxOldEncodeValueLen` ?

Reply:
new encoding value is always greater than 9

Reply:
You can refer to the comment in function CREATE